src/trie/merkle_tree.rs

```diff
@@ -63,23 +65,183 @@ impl ProofNode {
     }
 }
 
+pub(crate) struct MerkleTrees<H: StarkHash + Send + Sync, DB: BonsaiDatabase, CommitID: Id> {
+    pub db: KeyValueDB<DB, CommitID>,
+    _hasher: PhantomData<H>,
+    pub trees: HashMap<Vec<u8>, MerkleTree<H>>,
+}
+
+impl<H: StarkHash + Send + Sync, DB: BonsaiDatabase, CommitID: Id> MerkleTrees<H, DB, CommitID> {
+    pub(crate) fn new(db: KeyValueDB<DB, CommitID>) -> Self {
+        Self {
+            db,
+            _hasher: PhantomData,
+            trees: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn init_tree(
+        &mut self,
+        identifier: &[u8],
+    ) -> Result<(), BonsaiStorageError<DB::DatabaseError>> {
+        let tree = MerkleTree::new(&mut self.db, identifier.to_vec())?;
+        self.trees.insert(identifier.to_vec(), tree);
+        Ok(())
+    }
+
+    pub(crate) fn set(
+        &mut self,
+        identifier: &[u8],
+        key: &BitSlice<u8, Msb0>,
+        value: Felt,
+    ) -> Result<(), BonsaiStorageError<DB::DatabaseError>> {
+        let tree = self.trees.get_mut(identifier);
+        if let Some(tree) = tree {
+            tree.set(&mut self.db, key, value)
+        } else {
+            let mut tree = MerkleTree::new(&mut self.db, identifier.to_vec())?;
+            tree.set(&mut self.db, key, value)?;
+            self.trees.insert(identifier.to_vec(), tree);
+            Ok(())
+        }
+    }
+
+    pub(crate) fn get(
+        &self,
+        identifier: &[u8],
+        key: &BitSlice<u8, Msb0>,
+    ) -> Result<Option<Felt>, BonsaiStorageError<DB::DatabaseError>> {
+        let tree = self.trees.get(identifier);
+        if let Some(tree) = tree {
+            tree.get(&self.db, key)
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub(crate) fn contains(
+        &self,
+        identifier: &[u8],
+        key: &BitSlice<u8, Msb0>,
+    ) -> Result<bool, BonsaiStorageError<DB::DatabaseError>> {
+        let tree = self.trees.get(identifier);
+        if let Some(tree) = tree {
+            tree.contains(&self.db, key)
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub(crate) fn db_mut(&mut self) -> &mut KeyValueDB<DB, CommitID> {
+        &mut self.db
+    }
+
+    pub(crate) fn reset_to_last_commit(
+        &mut self,
+    ) -> Result<(), BonsaiStorageError<DB::DatabaseError>> {
+        for tree in self.trees.values_mut() {
+            tree.reset_to_last_commit(&mut self.db)?;
+        }
```

Comment:
Can you use rayon here to handle all trees in parallel?

Reply:
It's not really possible has the DB isn't shareable and nearly all the code of `reset_to_last_commit` is DB related. For the `commit` I split the compute of the hash and the DB insertion